infra/src/main/kotlin/fi/fta/geoviite/infra/tracklayout/LocationTrackService.kt

```diff
@@ -424,6 +425,51 @@ class LocationTrackService(
         return dao.fetchLocationTrackOwners()
     }
 
+    fun getSplittingInitializationParameters(
+        locationTrackId: IntId<LocationTrack>,
+        publishType: PublishType,
+    ): SplittingInitializationParameters {
+        logger.serviceCall(
+            "getSplittingInitializationParameters",
+            "locationTrackId" to locationTrackId,
+            "publishType" to publishType,
+        )
+        val locationTrack = getOrThrow(publishType, locationTrackId)
+        val switches = getSwitchesForLocationTrack(locationTrackId, publishType)
+            .mapNotNull { switchDao.fetchVersion(it, publishType) }
+            .map { switchDao.fetch(it) }
+            .mapNotNull { switch ->
+                switchLibraryService.getSwitchStructure(switch.switchStructureId).let { structure ->
+                    val presentationJointLocation = switch.getJoint(structure.presentationJointNumber)?.location
+                    if (presentationJointLocation != null) {
+                        switch to presentationJointLocation
+                    } else null
+                }
+            }
+            .map { (switch, location) ->
+                val address = geocodingService
+                    .getGeocodingContext(publishType, locationTrack.trackNumberId)
+                    ?.getAddressAndM(location)
+                SwitchOnLocationTrack(switch.id as IntId, switch.name, address?.address, location, address?.m)
+            }
+
+        val duplicateTracks = getLocationTrackDuplicates(locationTrackId, publishType).mapNotNull { duplicate ->
+            getWithAlignmentOrThrow(publishType, duplicate.id).let { (dupe, alignment) ->
+                geocodingService.getLocationTrackStartAndEnd(publishType, dupe, alignment)
+            }?.let { (start, end) ->
+                if (start != null && end != null) {
```

Comment:
Pienempi etu, mutta tämänkin iffin voisi korvata elviksellä jotenkin näin:
```
            getWithAlignmentOrThrow(publishType, duplicate.id)
              .let { (d, a) -> geocodingService.getLocationTrackStartAndEnd(publishType, d, a)    }
              ?.takeIf { (start, end) start != null && end != null }
              ?.let { (start, end) -> SplitDuplicateTrack(duplicate.id, duplicate.name, start, end) }
```

Reply:
Eikun jaa, voiko... enpäs olekaan varma suostuuko se tekemään autocastin non-null tyyppiin tuossa takeIf:ssä. Jos ei niin tietty meidän puolelta voisi olla template tyypitetty geneerinen `fun <T,S> takeIfValuesNotNull(pair: Pair<T?,S?>): Pair<T,S>?` jolla tuon voisi hoitaa letillä. Saattaisi olla jopa toteutettavissa pairille external funkkarina, mutta en ole varma.